test/tests/functional/pbs_hook_set_jobenv.py

```diff
@@ -0,0 +1,887 @@
+
+        # Read the output file and parse the values
+        self.env_nohook = {}
+        self.env_nohook_exclude = {}
+        self.penv = {}
+        self.read_env(job_outfile, "nohook")
+
+        # Now start the hooks
+        hook_name = "test_epi"
+        hook_body = """
+import pbs
+e = pbs.event()
+j = e.job
+j.Variable_List["happy"] = "days"
+pbs.logmsg(pbs.LOG_DEBUG,"Variable_List is %s" % (j.Variable_List,))
+"""
+
+        a2 = {'event': "execjob_epilogue", 'enabled': "true", 'debug': "true"}
+
+        self.server.create_import_hook(
+            hook_name,
+            a2,
+            hook_body,
+            overwrite=True)
+
+        # Submit a job with hooks in the system
+        jid2 = self.create_and_submit_job(attribs=a, content=script)
+        qstat = self.server.status(JOB, ATTR_o, id=jid2)
+        job_outfile = qstat[0][ATTR_o].split(':')[1]
+        self.server.expect(JOB, 'queue', op=UNSET, id=jid2, offset=10)
+
+        # read the output file for env with hooks
+        self.env_hook = {}
+        self.env_hook_exclude = {}
+        self.penv = {}
+        self.read_env(job_outfile, "hook")
+
+        # Validate
+        self.common_validate()
+
+        # Verify the env variables in logs too
+        self.common_log_match("mom")
+
+    def test_execjob_pro(self):
+        """
+        Test that environment variable not gets truncated
+        for execjob_prologue hook
+        """
+
+        a = {'Resource_List.select': '1:ncpus=1',
+             'Resource_List.walltime': 10}
+        script = []
+        script += ['#PBS -V']
+        script += ['env\n']
+        script += ['sleep 10\n']
+
+        # Submit a job without hooks in the system
+        jid = self.create_and_submit_job(attribs=a, content=script)
+        qstat = self.server.status(JOB, ATTR_o, id=jid)
+        job_outfile = qstat[0][ATTR_o].split(':')[1]
+        self.server.expect(JOB, 'queue', op=UNSET, id=jid, offset=10)
+
+        # read the output file for env without hook
+        self.env_nohook = {}
+        self.env_nohook_exclude = {}
+        self.penv = {}
+        self.read_env(job_outfile, "nohook")
+
+        # Now start the hooks
+        hook_name = "test_pro"
+        hook_body = """
+import pbs
+e = pbs.event()
+j = e.job
+j.Variable_List["happy"] = "days"
+pbs.logmsg(pbs.LOG_DEBUG,"Variable_List is %s" % (j.Variable_List,))
+"""
+        a2 = {'event': "execjob_prologue", 'enabled': "true", 'debug': "true"}
+
+        rv = self.server.create_import_hook(
+            hook_name,
+            a2,
+            hook_body,
+            overwrite=True)
+        self.assertTrue(rv)
+
+        # Submit a job with hooks in the system
+        jid2 = self.create_and_submit_job(attribs=a, content=script)
+        qstat = self.server.status(JOB, ATTR_o, id=jid2)
+        job_outfile = qstat[0][ATTR_o].split(':')[1]
+        self.server.expect(JOB, 'queue', op=UNSET, id=jid2, offset=10)
```

Comment:
For a prologue hook test, why does it matter that the job finishes?  Once it's in substate 42, the hook will have ran.

Reply:
We need to wait for the job to finish so that we can get and read the job output file, to validate environment variable settings (i.e. env output).